api/auth.go

```diff
@@ -126,3 +132,92 @@ func (a *API) writableOrganizationAddressesHandler(w http.ResponseWriter, r *htt
 	// write the response back to the user
 	apicommon.HTTPWriteJSON(w, userAddresses)
 }
+
+// oauthLoginHandler godoc
+//
+//	@Summary		Login using OAuth service
+//	@Description	Register/Authenticate a user and get a JWT token
+//	@Tags			auth
+//	@Accept			json
+//	@Produce		json
+//	@Param			request	body		apicommon.UserInfo	true	"Login credentials"
+//	@Success		200		{object}	apicommon.LoginResponse
+//	@Failure		400		{object}	errors.Error
+//	@Failure		401		{object}	errors.Error
+//	@Router			/oauth/login [post]
+func (a *API) oauthLoginHandler(w http.ResponseWriter, r *http.Request) {
+	// het the user info from the request body
+	loginInfo := &apicommon.OauthLoginRequest{}
+	if err := json.NewDecoder(r.Body).Decode(loginInfo); err != nil {
+		errors.ErrMalformedBody.Write(w)
+		return
+	}
+	log.Debugf("%v", loginInfo)
+	// get the user information from the database by email
+	user, err := a.db.UserByEmail(loginInfo.Email)
+	if err != nil && err != db.ErrNotFound {
+		errors.ErrGenericInternalServerError.Write(w)
+		return
+	}
+	// if the user doesn't exist, do oauth verifictaion and on success create the new user
+	if err == db.ErrNotFound {
+		// Register the user
+		// extract from the external signature the user pubkey and verify matches the provided one
+		if err := account.VerifySignature(loginInfo.OauthSignature, loginInfo.UserOauthSignature, loginInfo.Address); err != nil {
+			errors.ErrUnauthorized.WithErr(err).Write(w)
+			return
+		}
+		// fetch oauth service pubkey or address and verify the internal signature
+		resp, err := http.Get(fmt.Sprintf("%s/api/info/getAddress", a.oauthServiceURL))
+		if err != nil {
+			// TODO create new error for connection with the external service
+			errors.ErrGenericInternalServerError.WithErr(err).Write(w)
+			return
+		}
+		defer func() {
+			if err := resp.Body.Close(); err != nil {
+				// handle the error, for example log it
+				fmt.Println("Error closing response body:", err)
+			}
+		}()
+		var result apicommon.OauthServiceAddressResponse
+		if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
+			// TODO create new error for connection with the external service
+			errors.ErrGenericInternalServerError.WithErr(err).Write(w)
+			return
+		}
+
+		// verify the signature of the oauth service
+		if err := account.VerifySignature(loginInfo.Email, loginInfo.OauthSignature, result.Address); err != nil {
+			errors.ErrUnauthorized.WithErr(err).Write(w)
+			return
+		}
+
+		// genareate the new user and password and store it in the database
+		user = &db.User{
+			Email:     loginInfo.Email,
+			FirstName: loginInfo.FirstName,
+			LastName:  loginInfo.LastName,
+			Password:  internal.HexHashPassword(passwordSalt, loginInfo.UserOauthSignature),
+			Verified:  true,
+		}
+		if _, err := a.db.SetUser(user); err != nil {
+			errors.ErrGenericInternalServerError.WithErr(err).Write(w)
+		}
+	} else {
```

Comment:
no need for else clause, you can close the if here and unindent the next 6 lines.

another option if you want to handle more cases in the future, is replacing the "if, if, else" with a switch
```go
switch user, err := a.db.UserByEmail(loginInfo.Email); {
case err == db.ErrNotFound:
		// Register the user
		// extract from the external signature the user pubkey and verify matches the provided one
case err != nil:
	errors.ErrGenericInternalServerError.Write(w)
	return
default:
		// Login
		// check that the address generated password matches the one in the database
}
```

Reply:
Noted, no other cases are foreseen currently